### PR TITLE
Review Charm for Listing

### PR DESCRIPTION
**Pull request by kelkawi-a**
PR for reviewing charm code as defined in: https://discourse.charmhub.io/t/reviewing-charms/11698.

### CODE REVIEW

**Review by dimaqq — COMMENTED** [review]
### First pass review


Very high-level question, out of curiousity:

What's the motivation of charming Airtable specifically.

For example, there seem to be open-source alternatives...


Re Airtable specifically, I think there's a Helm Chart for that, how does this Charm compare to that?

File: src/scripts/pod-sweeper.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    # echo "Running kubectl command to get job pods..."
+    kubectl -n "${JOB_KUBE_NAMESPACE}" -L airbyte -l airbyte=job-pod \
+        get pods \
+        -o=jsonpath='{range .items[*]} {.metadata.name} {.status.phase} {.status.conditions[0].lastTransitionTime} {.status.startTime}{"\n"}{end}'
+}
+
+delete_pod() {
+    printf "From status '%s' since '%s', " "$2" "$3"
+    echo "$1" | grep -v "STATUS" | awk '{print $1}' | xargs --no-run-if-empty kubectl -n "${JOB_KUBE_NAMESPACE}" delete pod
+}
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/scripts/pod-sweeper.sh:18]: Airbyte itself creates k8s pods for different sync jobs. This script is meant to periodically clean up job pods which have exceeded a certain ttl value. I've added a note to the README to reflect that Airbyte creates these pods. 

Scaling with this charm has not been tested yet, but this might be modified depending on the outcome of such tests.

File: charmcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+
+    pod-successful-ttl-minutes:
+      description: Number of minutes until a successful job pod is removed.
+      default: 30
+      type: int
+
+    pod-unsuccessful-ttl-minutes:
+      description: Number of minutes until an unsuccessful job pod is removed.
+      default: 1440
+      type: int
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply charmcraft.yaml:416]: Airbyte by design is a complex application which exposes many configurations to limit resource usage on the k8s cluster. By default, the user can deploy the charm and leave all default configurations as is. These configurations are enabled to give flexibility to the user to add kube tolerations, node selectors and annotations. They are also here to have parity with the official Airbyte Helm chart as discussed with the Airbyte team.

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+
+
+"""Charm unit tests."""
+
+# pylint:disable=protected-access,too-many-public-methods
+
+import logging
+from unittest import TestCase, mock
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply tests/unit/test_charm.py:12]: Is this blocking? We currently use `unittest.TestCase` across all our team's charms, and this would require refactoring all the tests.

File: src/charm.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+            if not valid_pebble_plan:
+                logger.debug(f"failed to validate pebble plan for {container_name}, attempting creation again")
+                all_valid_plans = False
+                self._update(event)
+                continue
+
+            logger.info(f"performing up check for {container_name}")
+            check = container.get_check("up")
+            if check.status != CheckStatus.UP:
+                logger.error(f"check failed for {container_name}")
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/charm.py:176]: I tend to prefer more verbose code, given that there are multiple if checks being performed inside the for-loop.

File: src/charm.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+            if not CONTAINERS[container_name]:
+                continue
+
+            container = self.unit.get_container(container_name)
+            valid_pebble_plan = self._validate_pebble_plan(container, container_name)
+            logger.info(f"validating pebble plan for {container_name}")
+            if not valid_pebble_plan:
+                logger.debug(f"failed to validate pebble plan for {container_name}, attempting creation again")
+                all_valid_plans = False
+                self._update(event)
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/charm.py:170]: Good catch, i've updated this to only be called once if the `all_valid_plans` evaluates to `False`.

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+[testenv:fmt]
+description = Format the code
+deps =
+    black==22.8.0
+    isort==5.10.1
+commands =
+    isort {[vars]src_path} {[vars]tests_path}
+    black {[vars]src_path} {[vars]tests_path}
+
+[testenv:lint]
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply tox.ini:36]: I wasn't able to re-create this, perhaps a mismatch in tox versions?

```
git clone git@github.com:canonical/airbyte-k8s-operator.git
cd airbyte-k8s-operator
python -m venv venv
source venv/bin/activate
pip install -r dev-requirements.txt
tox
```

File: src/literals.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    "WORKSPACE_ROOT": "/workspace",
+    "CONFIG_ROOT": "/configs",
+    "CONFIGS_DATABASE_MINIMUM_FLYWAY_MIGRATION_VERSION": "0.35.15.001",
+    "JOBS_DATABASE_MINIMUM_FLYWAY_MIGRATION_VERSION": "0.29.15.001",
+    "MICRONAUT_ENVIRONMENTS": "control-plane",
+    "WORKERS_MICRONAUT_ENVIRONMENTS": "control-plane",
+    "CRON_MICRONAUT_ENVIRONMENTS": "control-plane",
+    "WORKLOAD_API_HOST": "localhost",
+    "MICROMETER_METRICS_ENABLED": "false",
+    "LAUNCHER_MICRONAUT_ENVIRONMENTS": "control-plane,oss",
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/literals.py:55]: Micronaut is used in Airbyte's [tech stack](https://docs.airbyte.com/understanding-airbyte/tech-stack), these environment variables are needed to configure it to the OSS deployment.

File: charmcraft.yaml [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+  airbyte-connector-builder-server:
+    resource: airbyte-connector-builder-server
+  airbyte-cron:
+    resource: airbyte-cron
+  airbyte-pod-sweeper:
+    resource: airbyte-pod-sweeper
+  airbyte-server:
+    resource: airbyte-server
+  airbyte-workers:
+    resource: airbyte-workers
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply charmcraft.yaml:436]: I did not find much guidance on the approach of using services vs. containers for a multi-container application. It might be possible to do this using services, but I am not sure how that would affect the charm operationally (e.g. observing logs).

We have not yet designed the scaling aspect of the Airbyte charms, but yes, the current assumption is that each juju unit/pod would contain these containers.

File: src/state.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+        Args:
+            app: workload application
+            get_relation: get peer relation method
+        """
+        # Use __dict__ to avoid calling __setattr__ and subsequent infinite recursion.
+        self.__dict__["_app"] = app
+        self.__dict__["_get_relation"] = get_relation
+
+    def __setattr__(self, name, value):
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/state.py:27]: This is actually being used across the charm for storing relation information (e.g. database and object storage credentials). An example of this can be seen under `relations/postgresql.py`.

File: src/state.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling charm state."""
+
+import json
+
+
+class State:
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/state.py:9]: Noted the typed wrapper as a future improvement on this wrapper class.

File: tests/unit/test_charm.py [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+                    "SYNC_JOB_RETRIES_COMPLETE_FAILURES_BACKOFF_BASE": 3,
+                    "SYNC_JOB_RETRIES_COMPLETE_FAILURES_BACKOFF_MAX_INTERVAL_S": 1800,
+                    "SYNC_JOB_RETRIES_COMPLETE_FAILURES_BACKOFF_MIN_INTERVAL_S": 10,
+                    "SYNC_JOB_RETRIES_COMPLETE_FAILURES_MAX_SUCCESSIVE": 5,
+                    "SYNC_JOB_RETRIES_COMPLETE_FAILURES_MAX_TOTAL": 10,
+                    "SYNC_JOB_RETRIES_PARTIAL_FAILURES_MAX_SUCCESSIVE": 1000,
+                    "SYNC_JOB_RETRIES_PARTIAL_FAILURES_MAX_TOTAL": 20,
+                    "TEMPORAL_HISTORY_RETENTION_IN_DAYS": 30,
+                    "TEMPORAL_HOST": "temporal-k8s:7233",
+                    "TEMPORAL_WORKER_PORTS": "9001,9002,9003,9004,9005,9006,9007,9008,9009,9010,9011,9012,9013,9014,9015,9016,9017,9018,9019,9020,9021,9022,9023,9024,9025,9026,9027,9028,9029,9030",
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply tests/unit/test_charm.py:389]: These are listed as an environment variable that is used by Airbyte. Reference [here](https://github.com/airbytehq/airbyte-platform/blob/275832edb013879ec475430fca7cef9c04854124/charts/airbyte/templates/env-configmap.yaml#L140).

**Comment by kelkawi-a** [comment]
> ### First pass review
> Very high-level question, out of curiousity:
> 
> What's the motivation of charming Airtable specifically.
> 
> For example, there seem to be open-source alternatives...
> 
> Re Airtable specifically, I think there's a Helm Chart for that, how does this Charm compare to that?

Thanks for the thorough review! The use of Airbyte is a business decision as its architecture matches what we are looking for in an ETL system, including the use of Temporal for durable execution of synchronization workflows.

This charm mostly has parity with the Helm chart in terms of the configs it exposes and the functionality it provides through the default deployment.

I've addressed most of the review comments in #16

File: charmcraft.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dimaqq) [reply charmcraft.yaml:416]: Well, this flies in the face of "The Model Driven Operator Manifesto"

https://juju.is/model-driven-operations-manifesto

> High-level config
> Operators should abstract complex application configuration. An operator should offer only high-level config to capture admin intent and reduce the burden of application specific domain expertise required to run the app.

P.S. I don't know the context behind this charm or who the airbyte team is, if possible please link some discussion.

My 2c: it's not a blocker, but exposing this many configuration knobs taxes administrators in the field.

File: charmcraft.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (dimaqq) [reply charmcraft.yaml:436]: Wrt. scaling, I figure it's fair game to take inspiration from the existing Helm chart.

File: src/relations/minio.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+                storage_data["namespace"],
+                storage_data["port"],
+                storage_data["secure"],
+            )
+
+            self.charm._state.minio = {
+                **storage_data,
+                "endpoint": endpoint,
+            }
+            self.charm._update(event)
```
Comment (dimaqq) [inline src/relations/minio.py:68]: Python doesn't enforce the distinction between public and private member/method names.

Then again this very PR creates this member and method.

Would it make sense to make them "public", i.e. `self.charm.state` and `self.charm.update()`?

File: src/state.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dimaqq) [reply src/state.py:27]: Thank you, found them. I'll leave comments there.

File: tests/unit/test_structured_config.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+def test_cpu_related_values(_harness) -> None:
+    """Test specific parameters for each field."""
+    erroneus_values = ["-123", "0", "100f"]
+    check_invalid_values(_harness, "job-main-container-cpu-limit", erroneus_values)
+    accepted_values = ["200m", "4"]
+    check_valid_values(_harness, "job-main-container-cpu-limit", accepted_values)
+
+
+def test_memory_related_values(_harness) -> None:
+    """Test specific parameters for each field."""
```
Comment (dimaqq) [inline tests/unit/test_structured_config.py:59]: Same docstring appearing in 3 different tests is not very useful.

**Review by dimaqq — COMMENTED** [review]
Here's another batch of comments.

Sorry that I'm unable to review the whole thing. I'm multi-tasking and the charm is kinda large.

File: tox.ini (hunk 6 above)
[the remark this answers is not in the document]
Comment (dimaqq) [reply tox.ini:36]: The issue here is that the tool is linting not only the charm code, but also the virtualenv that tox has created.

My host Python is 3.12 and `SafeConfigParser` was removed by that version. I would imagine that pflake8 has some kind of work-around included.

In short, it's not an issue for the "conventional" way of running things, but the command could be improved to scan only some files. This approach admittedly has its own problems.

If it works for you locally and in CI, keep it.

File: charmcraft.yaml [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+    __schema_source: https://raw.githubusercontent.com/canonical/operator-schemas/master/object-storage.yaml
+
+  s3-parameters:
+    interface: s3
+    limit: 1
+    optional: true
+
+provides:
+  airbyte-server:
+    interface: airbyte-server
```
Comment (dimaqq) [inline charmcraft.yaml:76]: Discussion on naming style: https://matrix.to/#/!xdClnUGkurzjxqiQcN:ubuntu.com/$oWHOH6C6yg_Ni0XpIq8X7zgGfLDgas9BEcie6xBIyMY?via=ubuntu.com&via=matrix.org&via=matrix.debian.social

**Review by dimaqq — CHANGES_REQUESTED** [review]
Please read https://juju.is/docs/sdk/holistic-vs-delta-charms

This applies to the airbyte-server relation event handling, where:
- if the relation is set up before the unit is ready, no data is published yet (as far as I can tell)
- charm relies on udpate-status event to detect that it's ready and publish relation data

The update status events happen every 5 minutes by default, however, the period is often set longer, possibly much longer in production. Which means that it may take a while for the application and relations to settle.

Additionally I'm not very happy to see `_provide_server_status` a private-ish method being called from another module.

Ideally, the charm would be event driven, where relation data may be ultimately published on some container's pebble ready event. The implementation would then likely look like a "holistic" charm.

Alternatively, you could set up a background process in a container, that checks stuff and issues a custom pebble notice when things change.

You could also chose to bite the bullet and keep the current architecture, even if it's a little hard to understand.

As a reviewer, my requirement is that there's an explicit choice.

**Comment by kelkawi-a** [comment]
> Please read https://juju.is/docs/sdk/holistic-vs-delta-charms
> 
> This applies to the airbyte-server relation event handling, where:
> 
> * if the relation is set up before the unit is ready, no data is published yet (as far as I can tell)
> * charm relies on udpate-status event to detect that it's ready and publish relation data
> 
> The update status events happen every 5 minutes by default, however, the period is often set longer, possibly much longer in production. Which means that it may take a while for the application and relations to settle.
> 
> Additionally I'm not very happy to see `_provide_server_status` a private-ish method being called from another module.
> 
> Ideally, the charm would be event driven, where relation data may be ultimately published on some container's pebble ready event. The implementation would then likely look like a "holistic" charm.
> 
> Alternatively, you could set up a background process in a container, that checks stuff and issues a custom pebble notice when things change.
> 
> You could also chose to bite the bullet and keep the current architecture, even if it's a little hard to understand.
> 
> As a reviewer, my requirement is that there's an explicit choice.

This architecture is being used on other charms developed by our team (e.g. [Temporal](https://github.com/canonical/temporal-k8s-operator), [Trino](https://github.com/canonical/trino-k8s-operator)), so i'd like to keep it as is for uniformity. There also doesn't seem to be a consensus around naming style for interfaces, some charms use kebab-case (e.g. nginx-ingress-integrator) for their interface.

The `airbyte-server` container is the charm container which handles Airbyte server logic, while the `airbyte-server` relation is the endpoint that the [Airbyte UI charm](https://github.com/canonical/airbyte-ui-k8s-operator) relates to to fetch information about the Airbyte server.

**Comment by dimaqq** [comment]
Minor comment to the above:
- Temporal charm does work in the same way indeed: https://github.com/canonical/temporal-k8s-operator/blob/45f4a36e5e0ec6a1190a14004103fa4e4da7f387/src/charm.py#L262-L287
- Trino charm does not: https://github.com/canonical/trino-k8s-operator/blob/4be540bebe5c28e77656607f51dd3fe064cf5acd/src/charm.py#L187-L195 (on_update_status does not trigger relation data update, it relies on event deferral instead)

**Comment by dimaqq** [comment]
We've discussed this in Charm Tech, and found the following:
- our general responsibility is to advise on good practices
- in rare cases (e.g. security or approach unlikely to work at all), we may block a charm
- the choice of architecture is the decision of your team
- improvement in the way each charm team structures their charms is viewed as an ongoing process that can't a specific charm review cannot be waiting for, rather it ought to be done out of band, through training, tooling or office hours

In summary, go ahead with your current approach and fixes staged in #16

**Review by dimaqq — APPROVED** [review]

File: README.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+To verify that the setup is running correctly, run
+`juju status --relations --watch 2s` and ensure that all pods are active and all
+required integrations exist.
+
+## Contributing
+
+This charm is still in active development. Please see the
+[Juju SDK docs](https://juju.is/docs/sdk) for guidelines on enhancements to this
+charm following best practice guidelines, and
+[CONTRIBUTING.md](./CONTRIBUTING.md) for developer guidance.
```
Comment (dimaqq) [inline README.md:103]: Please use a link with full hostname, so that the link works when this file is displayed on charmhub.

File: charmcraft.yaml [hunk 16]
```diff
@@ -0,0 +1,467 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+# This file configures Charmcraft.
+# See https://juju.is/docs/sdk/charmcraft-config for guidance.
+
+name: airbyte-k8s
+type: charm
+title: Airbyte Server
```
Comment (dimaqq) [inline charmcraft.yaml:9]: I would prefer if the CharmHub name and the README title matched.

Today. it's a little off.
Screenshot 2024-07-26 at 17 02 15

File: tests/unit/test_structured_config.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+import pytest
+from ops.testing import Harness
+
+from charm import AirbyteK8SOperatorCharm
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def _harness():
```
Comment (dimaqq) [inline tests/unit/test_structured_config.py:18]: Minor:

there are special rules for pytest fixtures.
because of that, there isn't really a "private" method notation.

```suggestion
def harness():
```

P.S. a fixture with the same name overrides a fixture from an outer scope, so even if there is a `harness` fixture somewhere, a custom name is not called for.

File: .github/workflows/publish_charm.yaml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+name: Publish to edge
+
+on:
+  push:
+    branches:
+      - main
```
Comment (dimaqq) [inline .github/workflows/publish_charm.yaml:6]: I must be a little confused... are tests run before the charm is published to the edge channel?

File: .github/workflows/test.yaml [hunk 19]
```diff
@@ -0,0 +1,9 @@
+name: Tests
+
+on:
+  pull_request:
```
Comment (dimaqq) [inline .github/workflows/test.yaml:4]: Unless that's implemented in some other way, I would recommend to also run tests on push to main.

The reason for that is if you've got 2 PR's, each may independently pass the tests, but then both are merged, the aggregate may still fail.

File: .github/workflows/publish_charm.yaml (hunk 18 above)
Comment (kelkawi-a) [reply .github/workflows/publish_charm.yaml:6]: The tests are run on the PR, once the tests pass, this workflow runs when the PR is merged with main to push the changes to the `edge` channel on Charmhub.

File: .github/workflows/test.yaml (hunk 19 above)
Comment (kelkawi-a) [reply .github/workflows/test.yaml:4]: This was the case previously with the [operator workflows](https://github.com/canonical/operator-workflows) developed by the IS Devops team, but was changed in favor of only running the tests on the PR. PRs will only be merged one at a time to ensure such a case does not happen.
